erocksdb using dirty nifs

benoitc opened the pull request:
This branch replace the custom thread pool by the usage of dirty nifs. It also contains the following changes:
- add support for column families
- refactor the code to reuse parts of it across functions.
- move the functions that need it under the erocksdb namespace and rename them to use the rocksdb style. Some functions have also been moved in different files to improve the readability.
- `put`, `get`and `delete` functions are now using `rocksdb:Put`,  `rocksdb:Get`,  `rocksdb:Delete` instead of using`rocksdb:Write` for all. 
- add in memory databases support
- use latest stable version of rocksdb
- add `erocksdb:flush/1` and `erocksdb:aproximate_size/3` functions
- add `rebar3` support

Let me know if you're interested by such change. I can also simply rename the procject as erocksdb_nif if you don't want to support such changes for now :)

benoitc committed:
add new options to improve write concurrency

- allow_concurrent_memtable_write: If true, allow multi-writers to update mem
  tables in parallel.

- enable_write_thread_adaptive_yield: If true, threads synchronizing with the
  write batch group leader will wait for up to write_thread_max_yield_usec before blocking on a mutex.

https://github.com/facebook/rocksdb/blob/master/include/rocksdb/options.h#L1223-L1244

benoitc committed:
don't run tests each time we build it

benoitc committed:
use the dirty nif feature instead of our own thread pool

rather tha using our own thred pool, reuse the dirty nif feature introduced in
latest version of erlang.

benoitc committed:
snapshot and release_snapshot are CPU bound

benoitc committed:
preliminary support for column famillies

this patch addd the following features:

- open a database with columns
- list columns in a database
- create a column
- delete a column

Columns are nif resources that live while the database resource exists.

TODO: add support for column in put/get/delete/write & iterators

benoitc committed:
add put/get/delete support for column families

benoitc committed:
add iterators/{3,4}

handle creation of iterators for column families

benoitc committed:
makes sur to close column iterrator when it's dropped.

benoitc committed:
update driver

benoitc committed:
split utils in utils.cc & utils.h

benoitc committed:
extract column_family code in its own file

benoitc committed:
move iterators handling to erocksdb_iter.cc

benoitc committed:
move snapshots operations in their own module

benoitc committed:
move kv operations to erocksdb_kv.(cc|h)

benoitc committed:
Merge branch 'barrel-alpha' of gitlab.com:barrel-db/erocksdb into barrel-alpha

benoitc committed:
add basho bench

benoitc committed:
add missing test file

benoitc committed:
use latests stable version of rocksdb

benoitc committed:
Merge branch 'dirty_nif' into barrel-alpha

benoitc committed:
add in memory databases support.

in memory databases are created using an empty string or "mem" for the db name

benoitc committed:
add erocksdb:flush/1 function

benoitc committed:
add get_approximate_size function

benoitc committed:
make sure to destroy the db

benoitc committed:
add the binary_to_slice helper

benoitc committed:
handle put and delete outside the iterator

benoitc committed:
add db test

benoitc committed:
make Get code more simpler

benoitc committed:
refactor code

- put all the api inside the erocksdb namespace
- add helpers where we can reuse the code
- use the rocksdb style to name functions

benoitc committed:
Merge branch 'simple-put-delete' into 'barrel-alpha'

refactor and GET/PUT improvements

This change add the following:
- simple `Put` and `Delete` function reusing `rocksdb::Db::PUT`and `rocksdb::Db::DELETE`
- function helpers to reuse code 
- refactor C code to use new helpers and new function namin in sync with rocksdb.

See merge request !2

benoitc committed:
compile with rebar3

benoitc committed:
fix tests on travis

mocchira commented:
@benoitc thanks for the great PR!
As reviewing this PR may take much time according to its size,
please wait for a while.
I will get in touch with you after finishing the review.

benoitc committed:
bump rocksdb to 4.5.1

benoitc committed:
compile against rebar3
T PUSH#

benoitc committed:
add get_latest_seq_number/1

allows to retrieve the latest sequence number

benoitc committed:
add basic features to allow the replication

This changes allows to iterate over a transaction log, retrieve the last
transaction as a binary and update another database with it. SO we can basically
replicate a database to another in a master-slave fashion.

benoitc committed:
Merge pull request #1 from barrel-db/expose_transactions_log

Expose transactions log

mocchira commented on c_src/build_deps.sh:
```diff
@@ -13,8 +13,7 @@ SCRIPT=$SCRIPTPATH/${0##*/}
 BASEDIR=$SCRIPTPATH
 BUILD_CONFIG=$BASEDIR/rocksdb/make_config.mk
 
```
Fix the url in the commit comment to reference NOT master BUT the specific commit ID

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1037,50 +920,45 @@ async_write(
+    }
 
     // Construct a write batch:
     rocksdb::WriteBatch* batch = new rocksdb::WriteBatch;
```
Use stack instead of heap. ( don't use `new` )

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1037,50 +920,45 @@ async_write(
+                enif_make_tuple2(env, ATOM_BAD_WRITE_ACTION, result));
     }   // if
 
     rocksdb::WriteOptions* opts = new rocksdb::WriteOptions;
```
Use stack instead of heap. ( don't use `new` )

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1037,50 +920,45 @@ async_write(
-    erocksdb::WorkTask* work_item = new erocksdb::WriteTask(env, caller_ref,
-                                                            db_ptr.get(), batch, opts);
+    delete batch;
+    delete opts;
```
no need to delete when using stack.

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1037,50 +920,45 @@ async_write(
+    const ERL_NIF_TERM& opts_ref   = argv[2];
 
     rocksdb::ReadOptions *opts = new rocksdb::ReadOptions();
 
```
Use stack instead of heap. ( don't use `new` )

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1095,42 +973,50 @@ async_get(
+    const bool keys_only = ((argc == 3) && (argv[2] == ATOM_KEYS_ONLY));
 
     rocksdb::ReadOptions *opts = new rocksdb::ReadOptions;
 
```
Use stack instead of heap. ( don't use `new` )

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1146,228 +1032,133 @@ async_iterator(
-    return ATOM_OK;
+    // release reference created during CreateItrObject()
+    enif_release_resource(itr_ptr);
+    opts=NULL;  // ptr ownership given to ItrObject
```
`opts = NULL;`

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1161,12 +1354,102 @@ checkpoint(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+        return enif_make_badarg(env);
+    }
+
+    char cf_name[4096];
```
strictly speaking,
as there is no limit about the length of a column family name,
we need to expand the `cf_name` dynamically according to the return value of `enif_get_string`.

so there are two options.
1. implement the above consideration
2. Add the todo comment about the above consideration

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1161,12 +1354,102 @@ checkpoint(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+parse_cf_descriptor(ErlNifEnv* env, ERL_NIF_TERM item,
+                    std::vector<rocksdb::ColumnFamilyDescriptor>& column_families)
+{
+    char cf_name[4096];
```
same with the above review

mocchira commented on src/erocksdb.erl:
```diff
@@ -194,18 +194,17 @@ init() ->
-    {error, not_implemeted}.
+open(Name, DbOpts, CfDescriptors) ->
+    open(Name, DbOpts ++ CfDescriptors).
+
```
Add the spec of `open_with_cf`

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1176,31 +1459,96 @@ erocksdb_open(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+            cf_list = enif_make_list_cell(env, cf, cf_list);
+        }
+    } catch (const std::exception& e) {
+        // pass through
```
why not return the error tuple with `e.what()` rather than `pass through`?

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1235,6 +1583,45 @@ erocksdb_close(
+            result = enif_make_list_cell(env, cf_name, result);
+        }
+    } catch (const std::exception& e) {
+        // pass through and return nullptr
```
why not return the error tuple with `e.what()` rather than `pass through`?

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1150,13 +1186,36 @@ get(
+    ReferencePtr<DbObject> db_ptr;
+    ReferencePtr<ColumnFamilyObject> cf_ptr;
 
     rocksdb::ReadOptions *opts = new rocksdb::ReadOptions();
```
Use stack instead of heap. ( don't use `new` )

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1324,6 +1324,70 @@ iterator(
+            enif_release_resource(itr_ptr);
+        }
+    } catch (const std::exception& e) {
+        // pass through and return nullptr
```
why not free `ItrObject`s created while for-looping and return the error tuple with `e.what()` 
rather than `pass through`?

benoitc commented:
@mocchira  thanks for the review! unfortunately the pr was a bit behibd the branch (github issue?). i just updated it. Unfortunately it didn' create a new pr as expeected sorry for that :(

Anyway, You will see that i've done a lot of refactoring. Some of your previous comments have already been fixed :) I will take care about the reste...

mocchira commented:
@benoitc 

> @mocchira thanks for the review! unfortunately the pr was a bit behibd the branch (github issue?). i just > updated it. Unfortunately it didn' create a new pr as expeected sorry for that :(

OK!
Understood.

> Anyway, You will see that i've done a lot of refactoring. Some of your previous comments have already > been fixed :) I will take care about the reste...

Thank you for the quick fixes!
Anyway, 
As I'm still on the half way for the review,
I may add the review comments later.
Please check it out.

mocchira commented on test/basho_bench_driver_rocksdb.erl:
```diff
@@ -77,10 +75,30 @@ run(get, KeyGen, _ValueGen, State) ->
     end;
 run(put, KeyGen, ValueGen, State) ->
-    case erocksdb:put(State#state.db_handle, KeyGen(), ValueGen(), []) of
+    print_status(State#state.ref, 1000),
```
Making `1000` configurable would be good.

mocchira commented on c_src/erocksdb.cc:
```diff
@@ -1119,6 +1120,38 @@ erocksdb_close(
+
+    if (!status.ok())
+    {
+        return error_tuple(env, erocksdb::ATOM_ERROR_DB_REPAIR, status);
```
why not add and use the flush specific error code rather than using the  `erocksdb::ATOM_ERROR_DB_REPAIR`?

mocchira commented on src/erocksdb.erl:
```diff
@@ -461,6 +462,12 @@ repair(_Name, _DBOpts) ->
+-spec get_approximate_size(db_handle(), binary(), binary(), boolean()) -> integer().
+get_approximate_size(_DbHandle, _SKey, _EKey, _IncludeMemtable) ->
+  erlang:nif_error({error, not_loaded}).
+
```
Please add the unit tests as follows.
1. check the return value whether it's `0` when there is no records between StartKey and EndKey.
2. check the return value whether it's `> 0` when there are some records between StartKey and EndKey.

mocchira commented on c_src/erocksdb.h:
```diff
@@ -62,8 +63,11 @@ namespace erocksdb {
+ERL_NIF_TERM Get(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]);
+ERL_NIF_TERM Put(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]);
+ERL_NIF_TERM Delete(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]);
 
```
As I found that you refactored function naming with `rocksdb` style (camel case) on the later commit and 
I agree with your refactoring, 
Pleae ignore this comment. (just left for the review log)

> As Camel Case is not suitable for the function naming on conventional cpp,
> please rename as follows.
> 
> ``` shell
> s/Write/write/
> s/Get/get/
> s/Put/put/
> s/Delete/delete/
> ```

mocchira commented on c_src/erocksdb_kv.cc:
```diff
@@ -321,4 +321,98 @@ get(
+        return enif_make_badarg(env);
+    }
+
+    rocksdb::WriteOptions* opts = new rocksdb::WriteOptions;
```
Use stack instead of heap.

mocchira commented on c_src/erocksdb_kv.cc:
```diff
@@ -321,4 +321,98 @@ get(
+        return enif_make_badarg(env);
+    }
+
+    rocksdb::WriteOptions* opts = new rocksdb::WriteOptions;
```
Use stack instead of heap.

mocchira commented on c_src/erocksdb_kv.cc:
```diff
@@ -231,89 +231,44 @@ Get(
         return enif_make_badarg(env);
     }
 
+    rocksdb::ReadOptions *opts = new rocksdb::ReadOptions();
```
Use stack instead of heap.

mocchira commented:
@benoitc I finally finished the whole review.
please check the comments. 
( the most important one is `Use stack instead of heap` found around constructing (Read|Write)Option. )

Though we're going to accept your whole PR after polishing as much as possible,
please remove the commit "add basic features to allow the replication" as this PR already has become very big.
so please PR that as the another one.

Thanks for the great contribution.

mocchira commented:
@benoitc 
As one test have failed on the travis with https://travis-ci.org/leo-project/erocksdb/builds/131532037 ,
Please also check it.

mocchira commented:
@benoitc 
As I merged the another PR,
Can you rebase this PR on the tip?

benoitc commented:
@mocchira i will do that sometimes next week. .I have been incredibly busy  these last days

mocchira commented:
@benoitc no problem. ping me anytime which is good for you.

mocchira commented:
@benoitc 

Hello. 
Could you tell me whether you are still working on this PR and having a plan to update?
I will close this PR if not.

Note: In future, we might take part of your commits into the code base as needed.

benoitc commented:
Yes you can close it. I have since gone to my own binding since the project wasn't moving that fast.

mocchira commented:
@benoitc Thanks for replying. Got it.
